src/include/ipc/stream.h

```diff
@@ -95,7 +95,9 @@ struct sof_ipc_stream_params {
 	uint32_t host_period_bytes;
 	uint16_t no_stream_position; /**< 1 means don't send stream position */
 
-	uint16_t reserved[3];
+	uint8_t codec_id;
```

Comment:
this needs a minor ABI update

Reply:
@plbossart what do you think of using the PCM_PARAMS IPC to send the API ID for compressed codecs? This field be ignored for PCM data.

Reply:
it's natural to think of this, set_params and hw_params are similar in nature, but the former takes extra information into account.

Your proposal to use the codec_id is too limiting in general. For example for AAC you will need to send the format information for ADTS or ADIF. It gets even worse if you ever want to support encoders.

````c
struct snd_codec {
	__u32 id;
	__u32 ch_in;
	__u32 ch_out;
	__u32 sample_rate;
	__u32 bit_rate;
	__u32 rate_control;
	__u32 profile;
	__u32 level;
	__u32 ch_mode;
	__u32 format;
	__u32 align;
	union snd_codec_options options;
	__u32 reserved[3];
} __attribute__((packed, aligned(4)));

````

Reply:
@plbossart we have the capability to send the format information through the kcontrol data. Is that enough?

Reply:
it will of course mean 2 IPC's instead of 1

Reply:
maybe, if the two opens below can be sorted out.

a) can we set the values from the kernel? I thought control values were set by userspace only.
b) can we guarantee that the additional information is available to the firmware when it receives this _params IPC?

Reply:
> a) can we set the values from the kernel? I thought control values were set by userspace only.
Yes, as long we have the means to locate the codec component ID during hw_params, we can send the values from the kernel. Something to look into.

> b) can we guarantee that the additional information is available to the firmware when it receives this _params IPC?
Yes, this can be guaranteed. The tricky part would be to identify the codec component during hw_params and send the COMP SET_DATA IPC before sending the PCM_PARAMS IPC

Reply:
thanks @ranj063, looks like you added more work on your own plate :-)

I kind of like this approach because to some extent all these codec parameters can be quite custom and it'd be a hassle to maintain this through an IPC where all the fields have to be described.

That said, we'd need some sort of ABI or signature to make sure the params are in the 'right' format.

Reply:
Maybe we should add new structure to populate the params that would ensure compatibility

Reply:
Well, on a) above and to be more specific, can we send the values from the kernel and *only the kernel* - without any userspace interference or ability to change stuff.

Reply:
> Well, on a) above and to be more specific, can we send the values from the kernel and _only the kernel_ - without any userspace interference or ability to change stuff.

Yes!

Reply:
do you have an example of how this might be done? I never understood the control layer beyond .put and .get I am afraid.

Reply:
> this needs a minor ABI update

Yes, I will update the ABI.

Reply:
@plbossart @ranj063 

> Maybe we should add new structure to populate the params that would ensure compatibility

That could be a good idea. If we could just find a way to embed `struct snd_codec` inside `struct sof_ipc_pcm_params` without breaking compatibility we can have a good start.

Reply:
But we don't have enough reserved bytes left to embed this whole thing. Maybe what we can do instead is send the and_codec as the data with comp.SET_DATA. Then there won't be compatibility issues as.well.

Reply:
And in that case, maybe we don't even need to modify the pcm_params ipc structure as snd_codec already has the api_id right?

Reply:
> And in that case, maybe we don't even need to modify the pcm_params ipc structure as snd_codec already has the api_id right?

Yes, but we need to modify pcm_params_ipc to add snd_codec inside it.

Reply:
> But we don't have enough reserved bytes left to embed this whole thing. Maybe what we can do instead is send the snd_codec as the data with comp.SET_DATA. Then there won't be compatibility issues as.well.

This could work in hw_params / set_params we could sent the normal SOF_IPC_STREAM_PCM_PARAMS IPC and then also call snd_sof_ipc_set_get_comp_data with SET_DATA containing the snd_codec. 

But is it possible to call snd_sof_ipc_set_get_comp_data from set_params/hw_params? For PCM case we can just use the default codec id (and maybe fix it later if Intel decides to add more codecs).

Reply:
but why? we can send it in the data part of the COMP SET_DATA IPC isnt it which will precede the PCM_PARAMS

Reply:
> But is it possible to call snd_sof_ipc_set_get_comp_data from set_params/hw_params? For PCM case we can just use the default codec id (and maybe fix it later if Intel decides to add more codecs).

Yes, absolutely

Reply:
we only need to do it from set_params() right? or do we need this in hw_params() too?